### PR TITLE
Improve cleanup of xml elements

### PR DESCRIPTION
**Pull request by bigfluffycookie**
fixes https://github.com/SonarSource/sonarlint-visualstudio/issues/4426

### CODE REVIEW

File: src/ConnectedMode.UnitTests/Migration/FileCleanerTestCases/Ruleset_IncludedTopLevel-project-key_Input.ruleset.xml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (bigfluffycookie) [inline src/ConnectedMode.UnitTests/Migration/FileCleanerTestCases/Ruleset_IncludedTopLevel-project-key_Input.ruleset.xml:1]: This test didnt need to be a ruleset I suppose, it could have been a project file aswell. It is just for testing that we do not remove the root element.

File: src/ConnectedMode/Migration/XmlFileCleaner.cs [hunk 2]
```diff
@@ -73,12 +73,35 @@ public string Clean(string content, LegacySettings legacySettings, CancellationT
             nodesToRemove.ForEach(node =>
             {
                 Debug.Assert(node.ParentNode != null);
-                node.ParentNode.RemoveChild(node);
+
+                // If the only child elements in the parent element are white spaces and the node the whole parent elemnt can be removed.
+                var currentNode = ShouldRemoveParentElement(node) ? node.ParentNode : node;
```
Comment (bigfluffycookie) [inline src/ConnectedMode/Migration/XmlFileCleaner.cs:78]: I am not sure currentNode makes much sense but I couldnt think of a better name.

File: src/ConnectedMode/Migration/XmlFileCleaner.cs [hunk 3]
```diff
@@ -73,12 +73,35 @@ public string Clean(string content, LegacySettings legacySettings, CancellationT
...
+
+                currentNode.ParentNode.RemoveChild(currentNode);
             });
 
             return xmlDocumentHelper.SaveToString(document);
         }
 
+        private static bool ShouldRemoveParentElement(XmlNode node)
+        {
+            foreach (XmlNode child in node.ParentNode.ChildNodes)
```
Comment (bigfluffycookie) [inline src/ConnectedMode/Migration/XmlFileCleaner.cs:94]: I wanted to do a cool linq expression here but it would not let me for some reason. ChildNodes is IEnumerable so it should have been possible but anyways, thats why we have a normal foreach loop now

File: src/ConnectedMode.UnitTests/Migration/FileCleanerTestCases/AdditionalFiles_XprojectkeyY_NoneItemGroup_Cleaned.xml [hunk 4]
```diff
@@ -7,18 +7,6 @@
...
-
-    
-  </ItemGroup>
-
-  <ItemGroup Label="Should be removed - VB">
-    
-
-    
-  </ItemGroup>
-
```
Comment (duncanp-sonar) [inline src/ConnectedMode.UnitTests/Migration/FileCleanerTestCases/AdditionalFiles_XprojectkeyY_NoneItemGroup_Cleaned.xml:21]: Nice!

File: src/ConnectedMode.UnitTests/Migration/FileCleanerTestCases/Ruleset_IncludedTopLevel-project-key_Input.ruleset.xml (hunk 1 above)
Comment (duncanp-sonar) [reply src/ConnectedMode.UnitTests/Migration/FileCleanerTestCases/Ruleset_IncludedTopLevel-project-key_Input.ruleset.xml:1]: The problem won't happen with project files as we never remove nodes that are directly parented under the root node. We only remove:

```
<Project>
    <PropertyGroup>
        <CodeAnalysisRuleSet> <-- could be removed
...
```

or 

```
<Project>
    <ItemGroup>
        <AdditionalFiles Include="...">   <-- could be removed
...
```

```

File: src/ConnectedMode/Migration/XmlFileCleaner.cs (hunk 2 above)
Comment (duncanp-sonar) [reply src/ConnectedMode/Migration/XmlFileCleaner.cs:78]: `nodeToRemove`? ;-)

File: src/ConnectedMode/Migration/XmlFileCleaner.cs [hunk 5]
```diff
@@ -73,12 +73,35 @@ public string Clean(string content, LegacySettings legacySettings, CancellationT
             nodesToRemove.ForEach(node =>
             {
                 Debug.Assert(node.ParentNode != null);
-                node.ParentNode.RemoveChild(node);
+
+                // If the only child elements in the parent element are white spaces and the node the whole parent elemnt can be removed.
+                var currentNode = ShouldRemoveParentElement(node) ? node.ParentNode : node;
+
+                // This will ensure that there is no white space left behind awhen removing the actual currentNode.
```
Comment (duncanp-sonar) [inline src/ConnectedMode/Migration/XmlFileCleaner.cs:80]: Typo - `awhen`

**Comment by duncanp-sonar** [comment]
It's probably adding a test with no whitespace between the parent node being removed and its siblings e.g.

```
<Project>
<ItemGroup>
<None Include="includeBefore"/><AdditionalFiles Include="sonarlint.xml">
<Link>..\sonarlint.xml</Link>
</AdditionalFiles><None Include="includeAfter/>
</ItemGroup>
</Project>
```
